grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java

```diff
@@ -17,98 +17,927 @@
-                                               .setPayload(request.getPayload())
-                                               .build());
-            responseObserver.onCompleted();
-        }
-    }
+    private static AtomicReference<HttpHeaders> CLIENT_HEADERS_CAPTURE = new AtomicReference<>();
+    private static AtomicReference<HttpHeaders> SERVER_TRAILERS_CAPTURE = new AtomicReference<>();
 
     @ClassRule
     public static ServerRule server = new ServerRule() {
         @Override
         protected void configure(ServerBuilder sb) throws Exception {
             sb.numWorkers(1);
             sb.port(0, HTTP);
+            sb.defaultMaxRequestLength(MAX_MESSAGE_SIZE);
 
             sb.serviceUnder("/", new GrpcServiceBuilder()
-                    .addService(new TestServiceImpl())
+                    .addService(new TestServiceImpl(Executors.newSingleThreadScheduledExecutor()))
+                    .setMaxInboundMessageSizeBytes(MAX_MESSAGE_SIZE)
+                    .setMaxOutboundMessageSizeBytes(MAX_MESSAGE_SIZE)
                     .enableUnframedRequests(true)
-                    .build());
+                    .build()
+                    .decorate(TestServiceImpl.EchoRequestHeadersInTrailers::new)
+                    .decorate((client, ctx, req) -> {
+                        CLIENT_HEADERS_CAPTURE.set(req.headers());
+                        HttpResponse res = client.serve(ctx, req);
+                        return new FilteredHttpResponse(res) {
+
+                            private boolean headersReceived;
+
+                            @Override
+                            protected HttpObject filter(HttpObject obj) {
+                                if (obj instanceof HttpHeaders) {
+                                    if (!headersReceived) {
+                                        headersReceived = true;
+                                    } else {
+                                        SERVER_TRAILERS_CAPTURE.set((HttpHeaders) obj);
+                                    }
+                                }
+                                return obj;
+                            }
+                        };
+                    }));
         }
     };
 
+    private TestServiceBlockingStub blockingStub;
+    private TestServiceStub asyncStub;
+
+    @Before
+    public void setUp() {
+        blockingStub = ClientFactory.DEFAULT
+                .newClient("gproto+" + server.httpUri("/"), TestServiceBlockingStub.class,
+                           ClientOption.DEFAULT_MAX_RESPONSE_LENGTH.newValue((long) MAX_MESSAGE_SIZE));
+        asyncStub = ClientFactory.DEFAULT
+                .newClient("gproto+" + server.httpUri("/"),
+                           TestServiceStub.class,
+                           // Streaming requests should not timeout mid-stream.
+                           // TODO(anuraag): See if timeout condition should be changed in HttpRequestSubscriber
```

Comment:
It seems that HttpRequestSubscriber will issue a write timeout for streaming requests. I guess the write timeout could be disabled for streaming requests like done here, or it could apply to each message within the request (the latter would add a lot of complication).